File: ED del Pedro/Ejercicios Haskell/Archivos Haskell/SetClient.hs
------------------------------------------------------------
-- Estructuras de Datos
-- Tema 3. Estructuras de Datos Lineales
-- Pablo López
--
-- Módulo cliente del TAD Conjunto
------------------------------------------------------------

module SetClient where

import           Data.Char                          (toUpper)
import           DataStructures.Set.SortedLinearSet

-- |
-- >>> mkSet [1..10]
-- SortedLinearSet(1,2,3,4,5,6,7,8,9,10)
-- mkSet "abracadabra"
-- SortedLinearSet('a','b','c','d','r')
mkSet :: Ord a => [a] -> Set a
mkSet xs = foldr insert empty xs

primos :: Set Integer
primos = mkSet [13, 7, 3, 2, 5, 11, 19]

-- |
-- >>> cardinal primos
-- 7
-- >>> cardinal (mkSet "abracadabra")
-- 5
cardinal :: Set a -> Integer
cardinal s = fold (\ _ n -> n + 1) 0 s

-- |
-- >>> suma primos
-- 60
suma :: Num a => Set a -> a
suma s = fold (+) 0 s

-- |
-- >>> setToList primos
-- [2,3,5,7,11,13,19]
setToList :: Set a -> [a]
setToList s = fold (:) [] s


-- |
-- >>> filtra even primos
-- SortedLinearSet(2)
-- >>> filtra (\ x -> x `mod` 5 == 0) (mkSet [1..100])
-- SortedLinearSet(5,10,15,20,25,30,35,40,45,50,55,60,65,70,75,80,85,90,95,100)

filtra :: Ord a => (a -> Bool) -> Set a -> Set a
filtra p s = fold f empty s
    where
      f x solResto
        | p x = insert x solResto
        | otherwise = solResto

-- |
-- >>> union' primos primos
-- SortedLinearSet(2,3,5,7,11,13,19)
-- >>> union primos (mkSet [1..10])
-- SortedLinearSet(1,2,3,4,5,6,7,8,9,10,11,13,19)
union' :: Ord a => Set a -> Set a -> Set a
union' a b = fold insert b a

-- |
-- >>> aplica (*2) primos
-- SortedLinearSet(4,6,10,14,22,26,38)
-- >>> aplica toUpper (mkSet "abracadabra")
-- SortedLinearSet('A','B','C','D','R')
aplica :: (Ord a, Ord b) => (a->b) -> Set a -> Set b
aplica f set = fold (\ x -> f x) empty set

-- |
-- >>>  maximo primos
-- Just 19
-- >>> maximo (mkSet "abracadabra")
-- Just 'r'
-- maximo empty
-- Nothing
maximo :: Ord a => Set a -> Maybe a
maximo = undefined


aplica :: (Ord a, Ord b) => (a->b) -> Set a -> Set b
aplica p s = fold f empty s
    where
      f x solResto
        | solResto = p x

File: ED del Pedro/ED/Java/IdeaProjects/Estructura de Datos/src/SetClient.java
/*
  Estructuras de Datos
  Grado en Ingeniería Informática, del Software y de Computadores
  Tema 3. TAD Lineales
  Pablo López
*/

import dataStructures.set.Set;
import dataStructures.set.SortedLinkedSet;

public class SetClient {

    public static void main(String[] args) {
        Set<Integer> s = new SortedLinkedSet<>();

        int[] xs = {35, 7, 1, 5, 7, 2, 52, 18, 0, 5};
        for (int x : xs) {  // iterador
            s.insert(x);
        }

        System.out.println(s);

        s.delete(7);
        s.delete(0);
        s.delete(52);
        s.delete(5);
        s.delete(100);
        s.delete(-1);
        s.delete(13);

        System.out.println(s);

        // uso del iterador

        // suma de elementos de un conjunto

        int sum = 0;
        for (int x : s) {
            sum += x;
        }
        System.out.println(sum);

        // máximo de un conjunto

        if (s.size() != 0) {
            int maximo = s.iterator().next();  // solo funciona si no está vacío
            for (int i : s) {
                if (maximo < i) {
                    maximo = i;
                }
            }
            System.out.println(maximo);
        }
    }
}
